packages/effect/src/ReadonlyRecord.ts

```diff
@@ -275,6 +275,44 @@ export const get: {
   <A>(self: ReadonlyRecord<A>, key: string): Option.Option<A> => has(self, key) ? Option.some(self[key]) : Option.none()
 )
 
+/**
+ * Apply a function to the element at the specified key, creating a new record,
+ * or return `None` if the key doesn't exist.
+ *
+ * @param self - The record to be updated.
+ * @param key - The key of the element to modify.
+ * @param f - The function to apply to the element.
+ *
+ * @example
+ * import { modify } from "effect/ReadonlyRecord"
+ * import { some, none } from "effect/Option"
+ *
+ * const f = (x: number) => x * 2
+ *
+ * assert.deepStrictEqual(
+ *  modify({ a: 3 }, 'a', f),
+ *  { a: 6 }
+ * )
+ * assert.deepStrictEqual(
+ *  modify({ a: 3 }, 'b', f),
+ *  { a: 3 }
+ * )
+ *
+ * @since 2.0.0
+ */
+export const modify: {
+  <A, B>(key: string, f: (a: A) => B): (self: ReadonlyRecord<A>) => Record<string, A | B>
+  <A, B>(self: ReadonlyRecord<A>, key: string, f: (a: A) => B): Record<string, A | B>
+} = dual(
+  3,
+  <A, B>(self: ReadonlyRecord<A>, key: string, f: (a: A) => B): Record<string, A | B> => {
+    if (!has(self, key)) {
+      return self
```

Comment:
should return `{ ...self }` as the return type is mutable

Reply:
The performance side of my brain is going crazy haha. But I guess it is either change the return type or return copies.

Reply:
I quite like the idea of a `.toRecord` function that converts a ReadonlyRecord to a Record by returning a copy.

Reply:
(you could then use ReadonlyRecord for the return types)

Reply:
I'm not sure, AFAIK returning a mutable type is the standard practice we follow (for example, in ReadonlyArray)

Reply:
Imho if one looks for perf those modules shouldn't be used, those are rather convenience utils. Returning ReadonlyRecord without copy is unsafe because the input may be mutable, readonly doesn't guarantee immutability it only says that something only performs read ops but it says nothing about the source (eg `Record<string, X>` can be assigned to `ReadonlyRecord<string, X>`)